### DISCUSSION TITLE
How can I learn and navigate the egs++ class library?

### DISCUSSION THREAD

**Discussion by jcampbell555** [nrc-cnrc/EGSnrc#793] [Questions] [answered]
I'm trying to experiment with making some new geometries and manipulating some of the example simulations in the EGS manual and the C++ class library but the library can be pretty hard to navigate. Is there any easy way to navigate these barriers or is it just sheer will and determination in learning each class? Is there an alphabetical list of all the libraries, classes, or commands? Any help would be appreciated. 
Thanks!

**Comment by rtownson** [comment, marked as answer]
Here is a list of geometries, not sure why is isn't alphabetical:
https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/group__Geometry.html

Here is a list of all the egs++ classes, which is alphabetical:
https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/annotated.html

Essentially you have to combine the "common" input syntax (https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/common.html) with the syntax you find for each of the sources, geometries, and ausgab objects on their respective pages.

Of course, let me know if you have any specific questions!

**Comment by jcampbell555** [comment]
@rtownson I know that egs_view uses levels of transparency to simulate energy of particles but is there a way to track the energy of a particle using a multitude of colors? Say for instance: 6 different colors for photons of energy 0-1 MeV, 1-2 MeV, 2-3 MeV, 3-4 MeV, 4-5 MeV, 5-6 MeV

**Reply by ftessier** [reply]
That would be a nice option in `egs_view`: color the particles by energy, say from blue to red, or to yellow on a viridis scale, between specified minimum and maximum (default to Emin to Emax).

**Reply by rtownson** [reply]
See the [track alpha code](https://github.com/nrc-cnrc/EGSnrc/blob/a6fc389c6465c949645380976f38a013a639759c/HEN_HOUSE/egs%2B%2B/view/egs_visualizer.cpp#L886-L888) in `HEN_HOUSE/egs++/view/egs_visualizer.cpp`:

```
if (energyScaling) {
    track_alpha = 0.2+0.8*image[idx].y;
}
```

Here, `image[idx].y` is the energy of the particle. So you could pretty easily add some lines to change the colour instead. You want to set `bCol` instead of `track_alpha`: it's an `EGS_Vector` of 3 RGB values. If you do this, of course feel free to share it or make a pull request.

**Comment by jcampbell555** [comment]
@rtownson I'm trying to create a gaussian beam that strikes a tungsten target but I'm having some trouble. I can simulate the tungsten target easily but the beam is more complicated. I know that the library used is egs_gaussian_shape but beyond that I'm unsure of the syntax. I've looked for examples but can't find any. Any help would be appreciated.

**Reply by rtownson** [reply]
You should be able to combine the gaussian shape with a parallel beam, for example.

https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/classEGS__ParallelBeam.html

https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/classEGS__GaussianShape.html

```
:start source:
  library = egs_parallel_beam
  name = some_name
  :start shape:
    library = egs_gaussian_shape
    :start shape:
      definition of the shape to be smeared
    :stop shape:
    sigma = 1 or 2 or 3 inputs
  :stop shape:
  :start spectrum:
    definition of the spectrum
  :stop spectrum:
  direction = Ux Uy Uz
  charge = -1 or 0 or 1 for electrons or photons or positrons
:stop source:
```

Where it says "definition of the shape to be smeared" you might put a circle shape, for example.

**Reply by jcampbell555** [reply]
could you combine gaussian shape with a collimated beam?

**Reply by rtownson** [reply]
Sure, anywhere it asks for a shape input block, you can use it. It just modifies the base shape that you give to the Gaussian shape.

**Reply by jcampbell555** [reply]
So here's what I input as my source:
:start source:
    library = egs_collimated_source
    name = collimated
    :start source shape:
        library = egs_gaussian_shape
	:start shape:
		library = egs_circle
		radius = 0.15
	:stop shape:
	sigma = 50
    :stop source shape:
    :start target shape:
        library   = egs_circle
        radius = 0.15
    :stop target shape:
    :start spectrum:
	type = monoenergetic
	energy = 20
    :stop spectrum:
    distance = 5
    charge = -1
:stop source:

And the error I receive in the GUI is:
   1EGS_CollimatedSource::getPositionDirection:
  my target shape circle_15, which is of type circle, failed to
  return a positive weight after 10000 attempts

**Reply by rtownson** [reply]
The source and target shapes are both located at the origin (by default). You'll need to move one of them, using a transformation inside the shape.

https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/classEGS__AffineTransform.html
(scroll to bottom for example)

**Reply by jcampbell555** [reply]
@rtownson is there a reason that my particles would only be displayed in a material and not entering or exiting the material?
![image](https://user-images.githubusercontent.com/87045590/144125829-6f6fbcd0-b57c-4ae1-a142-2768cc211560.png)

**Reply by rtownson** [reply]
If you want the particles to be transport in air (or vacuum) outside the geometry, you need to define it as a part of the geometry. E.g. use a big box surrounding everything. Otherwise, particles are discarded when they reach the edge of the defined geometry.

**Reply by jcampbell555** [reply]
@rtownson When reading the Summary of region dosimetry, what is the difference between "per particle" and "per fluence"?

**Reply by rtownson** [reply]
If 'last case' = 'fluence', then it will report 'per particle' because the fluence equals the number of primary source particles. For some source types (e.g. collimated source, radionuclide source..), these will not be equal so it reports per fluence instead.

**Reply by jcampbell555** [reply]
@rtownson so if I wanted to convert from per fluence to per particle, how would i do that?  Or rather if they are not equivalent, how is the total energy in the material calculated? Ex: if the fluence = 10, would I just multiply the Edep(per fluence) by 10?

**Reply by rtownson** [reply]
Which source are you using? You can multiply by the fluence and divide by `last case`, but per fluence is likely what you actually want. e.g. for the radionuclide source per fluence means per decay (or per disintegration) which is always what you want.

**Reply by ftessier** [reply]
One exception is the collimated source, where you might want to recover the result per emitted particle _in `4*pi`_, in which case you simply [divide the collimated source result by `4*pi`](https://egsnrc.slides.com/ftessier/egsnrc-collimated-source/#/1/2/5), assuming the source `distance` input is set to 1 (default).

**Reply by ftessier** [reply]
There is one exception: for a collimated source you might want to recover the results _per source particle in `4π`_. In that case, simply [divide the results by `4π`](https://egsnrc.slides.com/ftessier/egsnrc-collimated-source/#/1/2/5), assuming that the source `distance` parameter is set to 1 (the default). You need only multiply by the fluence for a single particle (1/4π for a unit sphere) to recover the result per particle!
